File: app/src/main/java/com/mgaetan89/showsrage/fragment/ShowsSectionFragment.java
package com.mgaetan89.showsrage.fragment;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v4.content.LocalBroadcastManager;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.mgaetan89.showsrage.Constants;
import com.mgaetan89.showsrage.R;
import com.mgaetan89.showsrage.adapter.ShowsAdapter;
import com.mgaetan89.showsrage.model.Show;
import com.mgaetan89.showsrage.model.ShowStat;
import com.mgaetan89.showsrage.model.ShowStatWrapper;
import com.mgaetan89.showsrage.model.ShowStats;
import com.mgaetan89.showsrage.model.ShowStatsWrapper;
import com.mgaetan89.showsrage.model.ShowsFilters;
import com.mgaetan89.showsrage.network.SickRageApi;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

public class ShowsSectionFragment extends Fragment implements View.OnClickListener {
	@Nullable
	private ShowsAdapter adapter = null;

	@Nullable
	private TextView emptyView = null;

	@NonNull
	private final List<Show> filteredShows = new ArrayList<>();

	@NonNull
	private final FilterReceiver receiver = new FilterReceiver(this);

	@Nullable
	private RecyclerView recyclerView = null;

	@NonNull
	private final Collection<Show> shows = new ArrayList<>();

	public ShowsSectionFragment() {
	}

	@Override
	public void onActivityCreated(Bundle savedInstanceState) {
		super.onActivityCreated(savedInstanceState);

		Bundle arguments = this.getArguments();

		if (arguments != null) {
			@SuppressWarnings("unchecked")
			Collection<Show> shows = (Collection<Show>) arguments.getSerializable(Constants.Bundle.INSTANCE.getSHOWS());

			if (shows != null) {
				this.filteredShows.addAll(shows);
				this.shows.addAll(shows);
			}
		}

		if (!this.shows.isEmpty()) {
			String command = getCommand(this.shows);
			Map<String, Integer> parameters = getCommandParameters(this.shows);

			SickRageApi.Companion.getInstance().getServices().getShowStats(command, parameters, new ShowStatsCallback(this));
		}

		this.updateLayout();

		if (this.adapter != null) {
			this.adapter.notifyDataSetChanged();
		}
	}

	@Override
	public void onClick(View view) {
		if (view == null) {
			return;
		}

		if (view.getId() == R.id.add_show) {
			AddShowFragment fragment = new AddShowFragment();
			View tabLayout = this.getActivity().findViewById(R.id.tabs);

			if (tabLayout != null) {
				tabLayout.setVisibility(View.GONE);
			}

			this.getActivity().getSupportFragmentManager().beginTransaction()
					.addToBackStack("add_show")
					.replace(R.id.content, fragment)
					.commit();
		}
	}

	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_shows_section, container, false);

		if (view != null) {
			this.emptyView = (TextView) view.findViewById(android.R.id.empty);
			this.recyclerView = (RecyclerView) view.findViewById(android.R.id.list);

			FloatingActionButton addShow = (FloatingActionButton) view.findViewById(R.id.add_show);

			if (addShow != null) {
				addShow.setOnClickListener(this);
			}

			if (this.recyclerView != null) {
				String showsListLayout = PreferenceManager.getDefaultSharedPreferences(this.getContext()).getString("display_shows_list_layout", "poster");
				int columnCount = this.getResources().getInteger(R.integer.shows_column_count);
				this.adapter = new ShowsAdapter(this.filteredShows, getAdapterLayoutResource(showsListLayout));

				this.recyclerView.setAdapter(this.adapter);
				this.recyclerView.setLayoutManager(new GridLayoutManager(this.getActivity(), columnCount));
			}
		}

		return view;
	}

	@Override
	public void onDestroy() {
		this.filteredShows.clear();
		this.shows.clear();

		super.onDestroy();
	}

	@Override
	public void onDestroyView() {
		this.emptyView = null;
		this.recyclerView = null;

		super.onDestroyView();
	}

	@Override
	public void onPause() {
		LocalBroadcastManager.getInstance(this.getContext()).unregisterReceiver(this.receiver);

		super.onPause();
	}

	@Override
	public void onResume() {
		super.onResume();

		IntentFilter intentFilter = new IntentFilter(Constants.Intents.INSTANCE.getACTION_FILTER_SHOWS());

		LocalBroadcastManager.getInstance(this.getContext()).registerReceiver(this.receiver, intentFilter);
	}

	@LayoutRes
	/* package */ static int getAdapterLayoutResource(String preferedLayout) {
		if ("banner".equals(preferedLayout)) {
			return R.layout.adapter_shows_list_content_banner;
		}

		return R.layout.adapter_shows_list_content_poster;
	}

	@NonNull
	/* package */ static String getCommand(@Nullable Iterable<Show> shows) {
		StringBuilder command = new StringBuilder();

		if (shows != null) {
			for (Show show : shows) {
				if (!isShowValid(show)) {
					continue;
				}

				if (command.length() > 0) {
					command.append("|");
				}

				command.append("show.stats_").append(show.getIndexerId());
			}
		}

		return command.toString();
	}

	@NonNull
	/* package */ static Map<String, Integer> getCommandParameters(@Nullable Iterable<Show> shows) {
		Map<String, Integer> parameters = new HashMap<>();

		if (shows != null) {
			for (Show show : shows) {
				if (!isShowValid(show)) {
					continue;
				}

				int indexerId = show.getIndexerId();

				parameters.put("show.stats_" + indexerId + ".indexerid", indexerId);
			}
		}

		return parameters;
	}

	/* package */
	static boolean isShowValid(@Nullable Show show) {
		return show != null && show.getIndexerId() > 0;
	}

	private void updateLayout() {
		if (this.filteredShows.isEmpty()) {
			if (this.emptyView != null) {
				this.emptyView.setVisibility(View.VISIBLE);
			}

			if (this.recyclerView != null) {
				this.recyclerView.setVisibility(View.GONE);
			}
		} else {
			if (this.emptyView != null) {
				this.emptyView.setVisibility(View.GONE);
			}

			if (this.recyclerView != null) {
				this.recyclerView.setVisibility(View.VISIBLE);
			}
		}
	}

	/* package */ static final class FilterReceiver extends BroadcastReceiver {
		private final WeakReference<ShowsSectionFragment> fragmentReference;

		FilterReceiver(ShowsSectionFragment fragment) {
			this.fragmentReference = new WeakReference<>(fragment);
		}

		@Override
		public void onReceive(Context context, Intent intent) {
			ShowsSectionFragment fragment = this.fragmentReference.get();

			if (fragment == null) {
				return;
			}

			SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
			String filterState = preferences.getString(Constants.Preferences.Fields.INSTANCE.getSHOW_FILTER_STATE(), Constants.Preferences.Defaults.INSTANCE.getSHOW_FILTER_STATE());
			int filterStatus = preferences.getInt(Constants.Preferences.Fields.INSTANCE.getSHOW_FILTER_STATUS(), Constants.Preferences.Defaults.INSTANCE.getSHOW_FILTER_STATUS());
			String searchQuery = intent.getStringExtra(Constants.Bundle.INSTANCE.getSEARCH_QUERY());

			Collection<Show> filteredShows = new ArrayList<>();
			Collection<Show> shows = fragment.shows;

			for (Show show : shows) {
				if (match(show, ShowsFilters.State.valueOf(filterState), filterStatus, searchQuery)) {
					filteredShows.add(show);
				}
			}

			fragment.filteredShows.clear();
			fragment.filteredShows.addAll(filteredShows);
			fragment.updateLayout();

			if (fragment.adapter != null) {
				fragment.adapter.notifyDataSetChanged();
			}
		}

		/* package */
		static boolean match(@Nullable Show show, ShowsFilters.State filterState, int filterStatus, String searchQuery) {
			return show != null &&
					matchFilterState(show, filterState) &&
					matchFilterStatus(show, filterStatus) &&
					matchSearchQuery(show, searchQuery);
		}

		/* package */
		static boolean matchFilterState(@NonNull Show show, @Nullable ShowsFilters.State filterState) {
			if (filterState != null) {
				switch (filterState) {
					case ACTIVE:
						return show.getPaused() == 0;

					case ALL:
						return true;

					case PAUSED:
						return show.getPaused() == 1;
				}
			}

			return false;
		}

		/* package */
		static boolean matchFilterStatus(@NonNull Show show, int filterStatus) {
			if (ShowsFilters.Status.Companion.isAll(filterStatus)) {
				return true;
			}

			String showStatus = show.getStatus();

			if (showStatus != null) {
				showStatus = showStatus.toLowerCase();

				switch (showStatus) {
					case "continuing":
						return ShowsFilters.Status.Companion.isContinuing(filterStatus);

					case "ended":
						return ShowsFilters.Status.Companion.isEnded(filterStatus);

					case "unknown":
						return ShowsFilters.Status.Companion.isUnknown(filterStatus);
				}
			}

			return false;
		}

		/* package */
		static boolean matchSearchQuery(@NonNull Show show, @Nullable String searchQuery) {
			if (searchQuery == null || searchQuery.isEmpty()) {
				return true;
			}

			searchQuery = searchQuery.trim();

			if (searchQuery.isEmpty()) {
				return true;
			}

			String showName = show.getShowName().toLowerCase();
			searchQuery = searchQuery.toLowerCase();

			return showName.contains(searchQuery);
		}
	}

	private static final class ShowStatsCallback implements Callback<ShowStatsWrapper> {
		private final WeakReference<ShowsSectionFragment> fragmentReference;

		ShowStatsCallback(ShowsSectionFragment fragment) {
			this.fragmentReference = new WeakReference<>(fragment);
		}

		@Override
		public void failure(RetrofitError error) {
			error.printStackTrace();
		}

		@Override
		public void success(ShowStatsWrapper showStatsWrapper, Response response) {
			ShowsSectionFragment fragment = this.fragmentReference.get();

			if (fragment == null) {
				return;
			}

			ShowStatWrapper data = showStatsWrapper.getData();

			if (data == null) {
				return;
			}

			Map<Integer, ShowStats> showStats = data.getShowStats();

			if (showStats != null) {
				for (Map.Entry<Integer, ShowStats> entry : showStats.entrySet()) {
					List<Show> filteredShows = fragment.filteredShows;
					ShowStat showStatsData = entry.getValue().getData();
					int indexerId = entry.getKey();

					for (int i = 0; i < filteredShows.size(); i++) {
						Show show = filteredShows.get(i);

						if (show.getIndexerId() == indexerId) {
							show.setEpisodesCount(showStatsData.getTotal());
							show.setDownloaded(showStatsData.getTotalDone());
							show.setSnatched(showStatsData.getTotalPending());

							if (fragment.adapter != null) {
								fragment.adapter.notifyItemChanged(i);
							}

							break;
						}
					}

					for (Show show : fragment.shows) {
						if (show.getIndexerId() == indexerId) {
							show.setEpisodesCount(showStatsData.getTotal());
							show.setDownloaded(showStatsData.getTotalDone());
							show.setSnatched(showStatsData.getTotalPending());

							break;
						}
					}
				}
			}

			fragment.updateLayout();
		}
	}
}

File: app/src/main/kotlin/com/mgaetan89/showsrage/fragment/ShowsSectionFragment.kt
package com.mgaetan89.showsrage.fragment

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.content.LocalBroadcastManager
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.GridLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.futuremind.recyclerviewfastscroll.viewprovider.DefaultScrollerViewProvider
import com.mgaetan89.showsrage.Constants
import com.mgaetan89.showsrage.R
import com.mgaetan89.showsrage.adapter.ShowsAdapter
import com.mgaetan89.showsrage.extension.getPreferences
import com.mgaetan89.showsrage.extension.getShows
import com.mgaetan89.showsrage.extension.getShowsFilterState
import com.mgaetan89.showsrage.extension.getShowsFilterStatus
import com.mgaetan89.showsrage.extension.getShowsListLayout
import com.mgaetan89.showsrage.extension.ignoreArticles
import com.mgaetan89.showsrage.extension.saveShowStat
import com.mgaetan89.showsrage.helper.Utils
import com.mgaetan89.showsrage.model.Show
import com.mgaetan89.showsrage.model.ShowStatsWrapper
import com.mgaetan89.showsrage.model.ShowsFilters
import com.mgaetan89.showsrage.network.SickRageApi
import io.realm.Realm
import io.realm.RealmChangeListener
import io.realm.RealmObject
import io.realm.RealmResults
import kotlinx.android.synthetic.main.fragment_shows_section.empty
import kotlinx.android.synthetic.main.fragment_shows_section.fastscroll
import kotlinx.android.synthetic.main.fragment_shows_section.list
import retrofit.Callback
import retrofit.RetrofitError
import retrofit.client.Response
import java.lang.ref.WeakReference
import java.util.Comparator

class ShowsSectionFragment : Fragment(), RealmChangeListener<RealmResults<Show>> {
	private val filteredShows = mutableListOf<Show>()
	private lateinit var realm: Realm
	private val receiver = FilterReceiver(this)
	private lateinit var shows: RealmResults<Show>
	private var swipeRefreshLayout: SwipeRefreshLayout? = null

	override fun onChange(shows: RealmResults<Show>) {
		if (!this.shows.isEmpty()) {
			val command = getCommand(this.shows)
			val parameters = getCommandParameters(this.shows)

			SickRageApi.instance.services?.getShowStats(command, parameters, ShowStatsCallback(this))
		}

		this.context?.let {
			val intent = Intent(Constants.Intents.ACTION_FILTER_SHOWS)

			LocalBroadcastManager.getInstance(it).sendBroadcast(intent)
		}
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)

		this.swipeRefreshLayout = this.activity?.findViewById(R.id.swipe_refresh) as SwipeRefreshLayout?
	}

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
			= inflater.inflate(R.layout.fragment_shows_section, container, false)

	override fun onDestroyView() {
		this.swipeRefreshLayout = null

		super.onDestroyView()
	}

	override fun onPause() {
		this.context?.let {
			LocalBroadcastManager.getInstance(it).unregisterReceiver(this.receiver)
		}

		super.onPause()
	}

	override fun onResume() {
		super.onResume()

		this.context?.let {
			val intentFilter = IntentFilter(Constants.Intents.ACTION_FILTER_SHOWS)

			LocalBroadcastManager.getInstance(it).registerReceiver(this.receiver, intentFilter)
		}
	}

	override fun onStart() {
		super.onStart()

		val arguments = this.arguments
		val anime = if (arguments?.containsKey(Constants.Bundle.ANIME) == true) {
			arguments.getBoolean(Constants.Bundle.ANIME)
		} else {
			null
		}

		this.realm = Realm.getDefaultInstance()
		this.shows = this.realm.getShows(anime, this)
	}

	override fun onStop() {
		if (this.shows.isValid) {
			this.shows.removeAllChangeListeners()
		}

		this.realm.close()

		super.onStop()
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		val preferences = view.context.getPreferences()
		val ignoreArticles = preferences.ignoreArticles()
		val showsListLayout = preferences.getShowsListLayout()
		val columnCount = this.resources.getInteger(R.integer.shows_column_count)

		this.list?.adapter = ShowsAdapter(this.filteredShows, showsListLayout, ignoreArticles)
		this.list?.layoutManager = GridLayoutManager(this.context, columnCount)

		this.fastscroll?.setViewProvider(object : DefaultScrollerViewProvider() {
			override fun onHandleGrabbed() {
				super.onHandleGrabbed()

				swipeRefreshLayout?.isEnabled = false
			}

			override fun onHandleReleased() {
				super.onHandleReleased()

				swipeRefreshLayout?.isEnabled = true
			}
		})

		this.fastscroll?.setRecyclerView(this.list)
	}

	private fun updateLayout() {
		if (this.filteredShows.isEmpty()) {
			this.empty?.visibility = View.VISIBLE
			this.list?.visibility = View.GONE
		} else {
			this.empty?.visibility = View.GONE
			this.list?.visibility = View.VISIBLE
		}
	}

	internal class FilterReceiver(fragment: ShowsSectionFragment) : BroadcastReceiver() {
		private val fragmentReference = WeakReference(fragment)

		override fun onReceive(context: Context?, intent: Intent?) {
			val fragment = this.fragmentReference.get() ?: return
			val preferences = context?.getPreferences() ?: return
			val filterState = preferences.getShowsFilterState()
			val filterStatus = preferences.getShowsFilterStatus()
			val ignoreArticles = preferences.ignoreArticles()
			val searchQuery = intent?.getStringExtra(Constants.Bundle.SEARCH_QUERY)
			val shows = fragment.shows
			val filteredShows = if (!shows.isValid || !shows.isLoaded) {
				emptyList()
			} else {
				shows.filter {
					match(it, filterState, filterStatus, searchQuery)
				}.sortedWith(Comparator<Show> { first, second ->
					val firstProperty = Utils.getSortableShowName(first, ignoreArticles)
					val secondProperty = Utils.getSortableShowName(second, ignoreArticles)

					firstProperty.compareTo(secondProperty)
				})
			}
			val currentFilteredShows = fragment.filteredShows.filter {
				RealmObject.isValid(it)
			}

			if (filteredShows != currentFilteredShows) {
				fragment.filteredShows.clear()
				fragment.filteredShows.addAll(filteredShows)
				fragment.list?.adapter?.notifyDataSetChanged()
			}

			fragment.updateLayout()
		}

		companion object {
			internal fun match(show: Show?, filterState: ShowsFilters.State?, filterStatus: Int, searchQuery: String?): Boolean {
				return show != null &&
						matchFilterState(show, filterState) &&
						matchFilterStatus(show, filterStatus) &&
						matchSearchQuery(show, searchQuery)
			}

			internal fun matchFilterState(show: Show, filterState: ShowsFilters.State?) = when (filterState) {
				ShowsFilters.State.ACTIVE -> show.paused == 0
				ShowsFilters.State.ALL -> true
				ShowsFilters.State.PAUSED -> show.paused == 1
				else -> false
			}

			internal fun matchFilterStatus(show: Show, filterStatus: Int): Boolean {
				if (ShowsFilters.Status.isAll(filterStatus)) {
					return true
				}

				val showStatus = show.status?.toLowerCase()

				return when (showStatus) {
					"continuing" -> ShowsFilters.Status.isContinuing(filterStatus)
					"ended" -> ShowsFilters.Status.isEnded(filterStatus)
					"unknown" -> ShowsFilters.Status.isUnknown(filterStatus)
					else -> false
				}
			}

			internal fun matchSearchQuery(show: Show, searchQuery: String?): Boolean {
				val query = searchQuery?.trim()

				if (query.isNullOrEmpty()) {
					return true
				}

				val showName = show.showName?.toLowerCase() ?: ""

				return showName.contains(query!!.toLowerCase())
			}
		}
	}

	private class ShowStatsCallback(fragment: ShowsSectionFragment) : Callback<ShowStatsWrapper> {
		private val fragmentReference = WeakReference(fragment)

		override fun failure(error: RetrofitError?) {
			error?.printStackTrace()
		}

		override fun success(showStatsWrapper: ShowStatsWrapper?, response: Response?) {
			val data = showStatsWrapper?.data ?: return
			val fragment = this.fragmentReference.get() ?: return
			val filteredShows = fragment.filteredShows
			val shows = fragment.shows
			val showStats = data.showStats

			showStats?.forEach {
				val indexerId = it.key
				val showStatsData = it.value.data
				val realmStat = if (showStatsData != null) {
					Realm.getDefaultInstance().use {
						it.saveShowStat(showStatsData, indexerId)
					}
				} else {
					null
				}

				filteredShows.filter { it.isValid && it.indexerId == indexerId }.forEachIndexed { i, show ->
					show.stat = realmStat ?: show.stat

					fragment.list?.adapter?.notifyItemChanged(i, Constants.Payloads.SHOWS_STATS)
				}

				if (shows.isValid) {
					shows.filter { it.isValid && it.indexerId == indexerId }.forEach {
						it.stat = realmStat ?: it.stat
					}
				}
			}
		}
	}

	companion object {
		internal fun getCommand(shows: Iterable<Show>?): String {
			val command = StringBuilder()

			shows?.filter { isShowValid(it) }?.forEach {
				if (command.isNotEmpty()) {
					command.append("|")
				}

				command.append("show.stats_").append(it.indexerId)
			}

			return command.toString()
		}

		internal fun getCommandParameters(shows: Iterable<Show>?): Map<String, Int> {
			val parameters = shows?.associate {
				if (isShowValid(it)) {
					val indexerId = it.indexerId

					"show.stats_$indexerId.indexerid" to indexerId
				} else {
					"" to 0
				}
			} ?: emptyMap()

			return parameters.filterKeys(String::isNotEmpty)
		}

		internal fun isShowValid(show: Show?) = show != null && show.indexerId > 0
	}
}
